experiments/distributedshell/src/main/scala/org/apache/gearpump/experiments/cluster/appmaster/AbstractAppMaster.scala

```diff
@@ -0,0 +1,142 @@
+      val actorToWorkerId = mutable.HashMap.empty[ActorRef, Int]
+      val groupedResource = allocations.groupBy(_.worker).mapValues(_.foldLeft(Resource.empty)((totalResource, request) => totalResource add request.resource)).toArray
+      allocations.foreach(allocation => actorToWorkerId.put(allocation.worker, allocation.workerId))
+
+      groupedResource.map((workerAndResources) => {
+        val (worker, resource) = workerAndResources
+        val executorContext = ExecutorContext(currentExecutorId, actorToWorkerId.get(worker).get, appId, self, resource)
+        val name = ActorUtil.actorNameForExecutor(appId, currentExecutorId)
+        val launcherName = s"launcher${currentExecutorId}"
+        val launcherPath = ActorUtil.getFullPath(context.system, self.path.child(launcherName))
+
+        val jvmSetting = Util.resolveJvmSetting(userConfig, systemConfig).executor
+        val launchJVM = ExecutorJVMConfig(jvmSetting.classPath, jvmSetting.vmargs,
+          classOf[ActorSystemBooter].getName, Array(name, launcherPath), appJar, username)
+
+        val launch = LaunchExecutor(appId, currentExecutorId, resource, launchJVM)
+
+        context.actorOf(Props(classOf[ExecutorLauncher], executorClass, worker, launch, executorContext, userConfig), launcherName)
+        currentExecutorId += 1
+      })
+  }
+
+  def workerMsgHandler : Receive = {
+    case RegisterExecutor(executor, executorId, resource, workerId) =>
+      //watch for executor termination
+      context.watch(executor)
+      def launchTask(remainResources: Resource): Unit = {
+        if (remainResources.greaterThan(Resource.empty)) {
+          val TaskLaunchData(taskClass, userConfig) = scheduleTaskOnWorker(workerId)
+
+          executor ! LaunchTask(TaskContext(executorId, appId, self), ActorUtil.loadClass(taskClass), userConfig)
+
+          //Todo: subtract the actual resource used by task
+          val usedResource = Resource(1)
+          launchTask(remainResources subtract usedResource)
+        }
+      }
+      launchTask(resource)
+  }
+
+  def selfMsgHandler : Receive = {
+    case LaunchExecutorActor(conf : Props, executorId : Int, daemon : ActorRef) =>
+      val executor = context.actorOf(conf, executorId.toString)
+      daemon ! BindLifeCycle(executor)
+  }
+
+  def scheduleTaskOnWorker(workerId: Int): TaskLaunchData
+
+  def repeatActionUtil(seconds: Int)(action : => Unit) : Cancellable = {
+    val cancelSend = context.system.scheduler.schedule(Duration.Zero, Duration(2, TimeUnit.SECONDS))(action)
+    val cancelSuicide = context.system.scheduler.scheduleOnce(FiniteDuration(seconds, TimeUnit.SECONDS), self, PoisonPill)
+    new Cancellable {
+      def cancel(): Boolean = {
+        val result1 = cancelSend.cancel()
+        val result2 = cancelSuicide.cancel()
+        result1 && result2
+      }
+
+      def isCancelled: Boolean = {
+        cancelSend.isCancelled && cancelSuicide.isCancelled
+      }
+    }
+  }
+}
```

Comment:
It should be easier to define the API into an AppMasters functionality. Right now adding a new message from Client to AppMaster is very involved since it must traverse master, appmanager.